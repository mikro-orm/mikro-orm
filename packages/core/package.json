{
  "name": "@mikro-orm/core",
  "version": "6.1.12",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "funding": "https://github.com/sponsors/b4nan",
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 18.12.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy && yarn run -T gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "yarn run -T rimraf ./dist",
    "compile": "yarn run -T tsc -p tsconfig.build.json",
    "copy": "node ../../scripts/copy.mjs"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "dataloader": "2.2.2",
    "dotenv": "16.4.5",
    "esprima": "4.0.1",
    "fs-extra": "11.2.0",
    "globby": "11.1.0",
    "mikro-orm": "6.1.12",
    "reflect-metadata": "0.2.2"
  }
}
