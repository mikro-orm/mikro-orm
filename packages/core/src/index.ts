/**
 * @packageDocumentation
 * @module core
 */
/* istanbul ignore file */
export {
  Constructor, Dictionary, PrimaryKeyType, PrimaryKeyProp, Primary, IPrimaryKey, FilterQuery, IWrappedEntity, EntityName, EntityData, Highlighter,
  AnyEntity, EntityProperty, EntityMetadata, QBFilterQuery, PopulateOptions, Populate, Loaded, New, LoadedReference, LoadedCollection,
  GetRepository, EntityRepositoryType, MigrationObject, DeepPartial, PrimaryProperty, Cast, IsUnknown, EntityDictionary, EntityDTO, PlainObject,
} from './typings';
export * from './enums';
export * from './errors';
export * from './exceptions';
export * from './MikroORM';
export * from './entity';
export * from './events';
export * from './EntityManager';
export * from './unit-of-work';
export * from './utils';
export * from './hydration';
export * from './drivers';
export * from './connections';
export * from './platforms';
export * from './types';
export * from './naming-strategy';
export * from './metadata';
export * from './cache';
export * from './decorators';
