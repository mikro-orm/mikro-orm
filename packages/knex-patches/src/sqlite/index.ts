export * from './SqliteKnexDialect';
export * from './SqliteTableCompiler';
