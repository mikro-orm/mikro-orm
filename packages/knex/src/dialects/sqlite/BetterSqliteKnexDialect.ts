import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class BetterSqliteKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  _driver() {
    return require('better-sqlite3');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
