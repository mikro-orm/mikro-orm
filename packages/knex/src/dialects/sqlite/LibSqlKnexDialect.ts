import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class LibSqlKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  get driverName() {
    return 'libsql';
  }

  _driver() {
    return require('libsql');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
