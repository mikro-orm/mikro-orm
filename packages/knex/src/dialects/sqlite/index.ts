export * from './BaseSqliteConnection';
export * from './BaseSqlitePlatform';
export * from './BaseSqliteSchemaHelper';
export * from './SqliteTableCompiler';
export * from './BetterSqliteKnexDialect';
export * from './LibSqlKnexDialect';
export * from './SqliteKnexDialect';
