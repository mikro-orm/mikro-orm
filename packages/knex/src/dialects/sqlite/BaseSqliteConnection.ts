import type { Knex } from 'knex';
import { ensureDir, readFile } from 'fs-extra';
import { dirname } from 'path';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';
import { Utils } from '@mikro-orm/core';

export abstract class BaseSqliteConnection extends AbstractSqlConnection {

  override async connect(): Promise<void> {
    this.createKnex();
    await ensureDir(dirname(this.config.get('dbName')!));
    await this.client.raw('pragma foreign_keys = on');
  }

  getDefaultClientUrl(): string {
    return '';
  }

  override getClientUrl(): string {
    return '';
  }

  override async loadFile(path: string): Promise<void> {
    const conn = await this.client.client.acquireConnection();
    await conn.exec((await readFile(path)).toString());
    await this.client.client.releaseConnection(conn);
  }

  protected override getKnexOptions(type: string): Knex.Config {
    return Utils.mergeConfig({
      client: type,
      connection: {
        filename: this.config.get('dbName'),
      },
      pool: this.config.get('pool'),
      useNullAsDefault: true,
    }, this.config.get('driverOptions'));
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'get') {
      return res[0];
    }

    if (method === 'all') {
      return res;
    }

    if (Array.isArray(res)) {
      return {
        insertId: res[res.length - 1]?.id ?? 0,
        affectedRows: res.length,
        row: res[0],
        rows: res,
      } as T;
    }

    return {
      insertId: res.lastInsertRowid,
      affectedRows: res.changes,
    } as unknown as T;
  }

}
