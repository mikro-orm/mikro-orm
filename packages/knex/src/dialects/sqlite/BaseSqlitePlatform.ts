import { Utils } from '@mikro-orm/core';
import { AbstractSqlPlatform } from '../../AbstractSqlPlatform';

export abstract class BaseSqlitePlatform extends AbstractSqlPlatform {

  override usesDefaultKeyword(): boolean {
    return false;
  }

  override usesReturningStatement(): boolean {
    return true;
  }

  override getCurrentTimestampSQL(length: number): string {
    return super.getCurrentTimestampSQL(0);
  }

  override getDateTimeTypeDeclarationSQL(column: { length: number }): string {
    return 'datetime';
  }

  override getEnumTypeDeclarationSQL(column: { items?: unknown[]; fieldNames: string[]; length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    if (column.items?.every(item => Utils.isString(item))) {
      return 'text';
    }

    /* istanbul ignore next */
    return this.getTinyIntTypeDeclarationSQL(column);
  }

  override getTinyIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getSmallIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getIntegerTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return 'integer';
  }

  override getFloatDeclarationSQL(): string {
    return 'real';
  }

  override getBooleanTypeDeclarationSQL(): string {
    return 'integer';
  }

  override getVarcharTypeDeclarationSQL(column: { length?: number }): string {
    return 'text';
  }

  override convertsJsonAutomatically(): boolean {
    return false;
  }

  override allowsComparingTuples() {
    return false;
  }

  /**
   * This is used to narrow the value of Date properties as they will be stored as timestamps in sqlite.
   * We use this method to convert Dates to timestamps when computing the changeset, so we have the right
   * data type in the payload as well as in original entity data. Without that, we would end up with diffs
   * including all Date properties, as we would be comparing Date object with timestamp.
   */
  override processDateProperty(value: unknown): string | number | Date {
    if (value instanceof Date) {
      return +value;
    }

    return value as number;
  }

  override getIndexName(tableName: string, columns: string[], type: 'index' | 'unique' | 'foreign' | 'primary' | 'sequence'): string {
    if (type === 'primary') {
      return this.getDefaultPrimaryName(tableName, columns);
    }

    return super.getIndexName(tableName, columns, type);
  }

  override getDefaultPrimaryName(tableName: string, columns: string[]): string {
    return 'primary';
  }

  override supportsDownMigrations(): boolean {
    return false;
  }

  override getFullTextWhereClause(): string {
    return `:column: match :query`;
  }

}
