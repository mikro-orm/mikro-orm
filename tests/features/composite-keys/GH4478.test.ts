import { Entity, PrimaryKey, ManyToOne, SimpleLogger, PrimaryKeyProp, wrap } from '@mikro-orm/core';
import { MikroORM } from '@mikro-orm/sqlite';
import { mockLogger } from '../../helpers';

@Entity()
class School {

  @PrimaryKey()
  schoolCode!: string;

}

@Entity()
class Class {

  [PrimaryKeyProp]?: ['school', 'academicYear', 'classCode'];

  @ManyToOne(() => School, { name: 'school_code', primary: true })
  school!: School;

  @PrimaryKey()
  academicYear!: string;

  @PrimaryKey()
  classCode!: string;

}

@Entity()
class StudentAllocation {

  [PrimaryKeyProp]?: ['studentId', 'academicYear'];

  @PrimaryKey()
  studentId!: string;

  @PrimaryKey()
  academicYear!: string;

  @ManyToOne(() => School, { name: 'school_code' })
  school!: School;

  @ManyToOne(() => Class, { fieldNames: ['school_code', 'academic_year', 'class_code'], nullable: true })
  class?: Class;

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [StudentAllocation],
    dbName: `:memory:`,
    loggerFactory: options => new SimpleLogger(options),
  });
  await orm.schema.createSchema();
});

afterAll(() => orm.close(true));

test(`GH issue 4478`, async () => {
  expect(await orm.schema.getCreateSchemaSQL()).toMatchSnapshot();

  const school = orm.em.create(School, { schoolCode: 'abc' });
  orm.em.create(StudentAllocation, { studentId: '1', academicYear: '2023', school });
  orm.em.create(StudentAllocation, {
    studentId: '2',
    academicYear: '2023',
    school,
    class: { school, classCode: 'cls', academicYear: '2023' },
  });

  const mock = mockLogger(orm);
  await orm.em.flush();
  expect(mock.mock.calls).toMatchSnapshot();
  orm.em.clear();

  const sa = await orm.em.find(StudentAllocation, { academicYear: '2023' }, { populate: ['*'] });
  expect(wrap(sa[0]).toObject()).toEqual({
    academicYear: '2023',
    school: {
      schoolCode: 'abc',
    },
    studentId: '1',
  });

  expect(wrap(sa[1]).toObject()).toEqual({
    academicYear: '2023',
    class: {
      academicYear: '2023',
      classCode: 'cls',
      school: {
        schoolCode: 'abc',
      },
    },
    school: {
      schoolCode: 'abc',
    },
    studentId: '2',
  });
});
