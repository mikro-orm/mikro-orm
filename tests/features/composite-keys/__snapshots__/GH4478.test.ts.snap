// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GH issue 4478 1`] = `
"pragma foreign_keys = off;

create table \`school\` (\`school_code\` text not null, primary key (\`school_code\`));

create table \`class\` (\`school_code\` text not null, \`academic_year\` text not null, \`class_code\` text not null, constraint \`class_school_code_foreign\` foreign key(\`school_code\`) references \`school\`(\`school_code\`) on update cascade, primary key (\`school_code\`, \`academic_year\`, \`class_code\`));
create index \`class_school_code_index\` on \`class\` (\`school_code\`);

create table \`student_allocation\` (\`student_id\` text not null, \`academic_year\` text not null, \`school_code\` text not null, \`class_code\` text null, constraint \`student_allocation_school_code_foreign\` foreign key(\`school_code\`) references \`school\`(\`school_code\`) on update cascade, constraint \`student_allocation_school_code_academic_year_class_code_foreign\` foreign key(\`school_code\`, \`academic_year\`, \`class_code\`) references \`class\`(\`school_code\`, \`academic_year\`, \`class_code\`) on delete set null on update cascade, primary key (\`student_id\`, \`academic_year\`));
create index \`student_allocation_school_code_index\` on \`student_allocation\` (\`school_code\`);
create index \`student_allocation_school_code_academic_year_class_code_index\` on \`student_allocation\` (\`school_code\`, \`academic_year\`, \`class_code\`);

pragma foreign_keys = on;
"
`;

exports[`GH issue 4478 2`] = `
[
  [
    "[query] begin",
  ],
  [
    "[query] insert into \`school\` (\`school_code\`) values ('abc')",
  ],
  [
    "[query] insert into \`class\` (\`school_code\`, \`academic_year\`, \`class_code\`) values ('abc', '2023', 'cls')",
  ],
  [
    "[query] insert into \`student_allocation\` (\`student_id\`, \`academic_year\`, \`school_code\`, \`class_code\`) values ('1', '2023', 'abc', NULL), ('2', '2023', 'abc', 'cls')",
  ],
  [
    "[query] commit",
  ],
]
`;
